backend/infrahub/auth.py

```
@@ -256,3 +266,144 @@ async def get_groups_from_provider(
 
     return []
+
+
+def safe_get_response_body(response: httpx.Response) -> str | dict[str, Any]:
+    """Safely extract response body from HTTP response.
+
+    Args:
+        response: The HTTP response object
+
+    Returns:
+        The response body as JSON dict if possible, otherwise as text, or error message
+    """
+    try:
+        return response.json()
+    except Exception:
+        try:
+            return response.text
+        except Exception:
+            return "Unable to parse response body"
+
+
+def extract_auth_error_message(response_body: str | dict[str, Any], base_message: str) -> str:
+    """Extract error message from OAuth 2.0/OIDC provider response following RFC 6749.
+
+    Args:
+        response_body: The response body from the authentication provider
+        base_message: Base error message to use if no specific error is found
+
+    Returns:
+        Formatted error message with provider details if available
+    """
+    if not isinstance(response_body, dict):
+        return base_message
+
+    # RFC 6749 standard error response format
+    error_description = response_body.get("error_description")
+    error_code = response_body.get("error")
+
+    if error_description:
+        return f"{base_message}: {error_description}"
+    if error_code:
+        return f"{base_message}: {error_code}"
+
+    return base_message
+
+
+def validate_auth_response(response: httpx.Response, provider_type: str = "authentication") -> None:
+    """Validate HTTP response from OAuth 2.0/OIDC provider and raise appropriate errors.
+
+    Args:
+        response: The HTTP response from the authentication provider
+        provider_type: Type of provider for logging (e.g., "OAuth 2.0", "OIDC")
+
+    Raises:
+        GatewayError: When the response indicates an error or invalid state
+    """
+    if 200 <= response.status_code <= 299:
+        return
```

Before we return I think we want to also validate that we can do a .json() on the response. I.e. it could be some incorrect url that returns a 200 status but still not be the one that we want.

---

Adjusted, response body parsing should raise an exception if response text is empty.

---

Need to review it, now it shortcuts everything when message is empty, which is the case for 404 for instance

---

To tweak it a bit further we could ensure that we can still get a json response. It looks like we might validate this but that `safe_get_response_body()` could return a plain string we got a 200 response that didn't contain any json data.

---

I think this is an improvement as it is but `safe_get_response_body()` would still be happy if there was a non json response that returned data. I.e. if the response was something like:

```html
<!doctype html><div>Loading…</div>
```